Update Applicant Information Form; Add Review Box

nathanielbuck opened the pull request:
## Description
Add the Date of Birth field and remove the SSN field. Additionally,
update the form title and subtitle.

This also adds a new form paradigm: the Review Box (for lack of a
better name). This is meant to display pre-populated data for the
veteran to review with the option to edit the information. The
primary purpose of this new read-first, edit-later paradigm is to
provide a better review experience and discourage easy editing for
fields that most likely shouldn't change, like name and birthday.

## Original issue(s)


## Testing done


## Screenshots


## Acceptance criteria
- [ ]

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

cesaresantiago commented:
Can we get a code review from @bkjohnson  or @cvalarida ? 

cc: @nathanielbuck

timwright12 commented on src/applications/my-education-benefits/containers/App.jsx:
```diff
@@ -1,12 +1,22 @@
-      {children}
-    </RoutedSavableApp>
+    <>
+      <Breadcrumbs>
```
Can we fill these out with the href values?

nathanielbuck replied:
I'm not entirely sure if we have these values at the moment.

timwright12 replied:
(syncd in slack)

timwright12 commented on src/applications/my-education-benefits/containers/CustomReviewDOBField.jsx:
```diff
@@ -0,0 +1,11 @@
+
+export default function CustomReviewDOBField({ children, uiSchema }) {
+  return (
+    <div className="review-row">
```
The parent of a dt/dd needs to be a `<dl>`

nathanielbuck replied:
Whoops! Not sure how this happened.

timwright12 commented on src/platform/forms/sass/_m-review-box-field.scss:
```diff
@@ -0,0 +1,42 @@
+.review-box {
```
Can this Sass be scoped in the application directory? Ideally we want to isolate these kinds of things

nathanielbuck replied:
Okay. Should I also be moving ReviewBoxField.jsx to be in the application directory?

timwright12 replied:
If you can, yeah. I'm not super familiar with the forms system (cc @cvalarida), but we try and scope all the code for an individual app within that app if possible

nathanielbuck replied:
Okay. I think this field may eventually be used other places, but we can cross that bridge when we get there.

cvalarida commented on src/platform/forms-system/src/js/components/FieldTemplate.jsx:
```diff
@@ -47,7 +47,11 @@ export default function FieldTemplate(props) {
   let errorClass;
-  if (hasErrors) {
+
+  if (
+    hasErrors &&
+    !(uiSchema && uiSchema['ui:options'] && uiSchema['ui:options'].hideErrors)
+  ) {
```
This seems like a bad idea. Why would we want to hide the errors?

nathanielbuck replied:
For the markup hierarchy to work with the date widget within the review box (to display correctly), I had to manually hide the errors here and display them in the review box. This was only necessary for the date widget; the markup and errors displayed fine for the full name field. For how I displayed the errors within the Review Box, see ReviewBoxField.jsx, lines 437-464.

And I'm definitely open to an alternative implementation that would require this extra step.

cvalarida replied:
Commenting here for posterity: @nathanielbuck and I met and talked about it. What we did to fix the problem he was running into was to wrap the `dateOfBirth` in a `view:dateOfBirth` field and apply the `ReviewBoxField` to that parent object instead of directly on the `dateOfBirth`.

cvalarida commented on src/platform/forms-system/src/js/components/FieldTemplate.jsx:
```diff
@@ -72,10 +76,13 @@ export default function FieldTemplate(props) {
+      (schema.type === 'object' ||
+        schema.type === 'array' ||
+        (schema.type === 'boolean' && !uiSchema['ui:widget']))) ||
+    (uiSchema &&
+      uiSchema['ui:options'] &&
+      uiSchema['ui:options'].noWrapperContent);
```
Why does the existing functionality not work for you?

nathanielbuck replied:
Similar to my comment above, things were getting messed up when the date widget was within my custom field.

cvalarida commented on src/platform/forms-system/src/js/components/FormNav.jsx:
```diff
@@ -97,6 +105,8 @@ export default function FormNav(props) {
           {!showHeader && (
             <div className="vads-u-font-size--h4">{stepText}</div>
           )}
+          {subTitle && <h3>{subTitle}</h3>}
+          {instructions && <p>{instructions}</p>}
```
Looks like what you want to do for your form is create a [block of text on the page](https://department-of-veterans-affairs.github.io/veteran-facing-services-tools/forms/common-patterns-for-building-forms#creating-a-block-of-text-with-no-associated-fields), right?

nathanielbuck replied:
Oh, yes. I will try that. Thanks for pointing me to the documentation.

cvalarida reviewed (commented):
Can you help me understand why the forms library needs to be modified here?

In the future, please pull the forms library changes into their own PR to aid in review. You can find our guidelines for submitting pull requests on our [platform documentation site](https://depo-platform-documentation.scrollhelp.site/developer-docs/Submitting-pull-requests-for-approval.655032351.html).

cvalarida commented:
What do you gain by adding the `ReviewBoxField` to your application that the [`ReviewCardField`](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/platform/forms-system/src/js/components/ReviewCardField.jsx) doesn't provide?

nathanielbuck commented:
> What do you gain by adding the `ReviewBoxField` to your application that the [`ReviewCardField`](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/platform/forms-system/src/js/components/ReviewCardField.jsx) doesn't provide?

Honestly, mostly styling changes. I wasn't sure how to cleanly reuse `ReviewBoxField` to meet our needs without potentially breaking it for others. I had to make a fair bit of changes to get it to work with the date widget.

nathanielbuck committed:
Update Applicant Information Form; Add Review Box

Add the Date of Birth field and remove the SSN field. Additionally,
update the form title and subtitle.

This also adds a new form paradigm: the Review Box (for lack of a
better name). This is meant to display pre-populated data for the
veteran to review with the option to edit the information. The
primary purpose of this new read-first, edit-later paradigm is to
provide a better review experience and discourage easy editing for
fields that most likely shouldn't change, like name and birthday.

nathanielbuck committed:
Add Additional Form Validation
